0000-access-keys.md

```diff
@@ -0,0 +1,313 @@
+- Proposal Code Name: access_keys
+- Start Date: 2019-07-08
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+- Issue(s): [nearprotocol/nearcore#687](https://github.com/nearprotocol/nearcore/issues/687)
+
+# Summary
+[summary]: #summary
+
+Access keys provide limited access to an account.
+Each access key belongs to some account and identified by a unique (within the account) public key.
+One account may have large number of access keys.
+Access keys will replace original account-level public keys.
+Access keys allow to act on behalf of the account by restricting allowed transactions with the access key permissions.
+
+# Motivation
+[motivation]: #motivation
+
+Access keys give an ability to use dApps in a secure way without asking the user to sign every transaction in the wallet.
+By issuing the access key once for the application, the application can now act on behalf of the user in a restricted environment.
+This enables seamless experience for the user.
+
+Access keys also enable a few other use-cases that are discussed in details below.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Here are proposed changes for the AccessKey and Account structs.  
+
+```rust
+/// `account_id,public_key` is a key in the state
+struct AccessKey {
+  /// The nonce for this.
+  pub nonce: Nonce,  // u64 
+  
+  /// Defines permissions for the AccessKey 
+  pub permission: AccessKeyPermission,
+}
+
+/// Defines permissions for AccessKey 
+pub enum AccessKeyPermission {
+  /// Restricts AccessKey to only be used for function calls.
+  FunctionCall(FunctionCallPermission),
+
+  /// Gives full access to the account.
+  /// NOTE: It's used to replace account-level public keys.
+  FullAccess,
+}
+
+pub struct FunctionCallPermission {
+  /// The amount that can be spent for transaction fees by this access key from the account balance.
+  /// When used, both account balance and the allowance is decreased.
+  pub allowance: Balance,  // u128
+
+  /// The AccountID of the receiver of the transaction. The access key will restrict transactions to
+  /// only this receiver.
+  pub receiver_id: AccountId,  // String
+  
+  /// If not None, the access key would be restricted to call only the given method name.
+  pub method_name: Option<String>,
```

Comment:
What if it's `None`? Will it be allowed to call any method on the contract? I thought it's what `FullAccess` is for.

Reply:
None would mean key is not restricted to a single method name, but can call any method name on the given `receiver_id` contract.